src/CSE.cpp

```diff
@@ -161,15 +161,25 @@ class GVN : public IRMutator {
         Expr value = mutate(let->value);
 
         // Make references to the variable point to the value instead.
-        let_substitutions.push(let->name, number);
+        // This shouldn't be done if the Let's value is a Load and occurs
+        // within a protected scope, since the value is not added to the
+        // numbering
+        if (value.as<Load>() == nullptr || !protect_loads_in_scope) {
```

Comment:
I would find this easier to parse as !(value.as<Load> && protect_loads_in_scope)